old_modules/m_ncr3000.man
.SH "NCR System 3000 CREDITS"
The SVR4 port was initially written by Andrew Herbert.  He was guided by a SVR4
port of top version 2.1 which was done by Andy Crump (andyc@bucky.intel.com).
Robert Boucher (boucher@sofkin.ca) adapted it to top version 3.1.
Jeff Janvrin (jeff.janvrin@columbiasc.ncr.com) ported version 3.3 to
the /stats filesystem


Changes
Wed Apr 18 2007 - wnl (3.6.1)
	Fixed a few bugs in sigconv.awk that were causing incorrect
	results on FreeBSD.  Changed configure.ac to fix a few linux
	problems: signal include file and /proc/1/stat.

Fri Apr 13 2007 - wnl (3.6.1)
	Removed the use of VPATH for compiling the system module and used
	an explicit dependency in the Makefile instead.  VPATH is now set
	to just srcdir to ensure that top will compile correctly when
	configured from a different directory.  On systems without VPATH
	support, top will still configure and compile, but only
	from within the source directory.  This fixes bug 1699526.

Fri Feb  2 2007 - wnl (3.6.1)
	Revised the way that configure figures out owner, group, and mode.
	For systems that don't use the kernel, it tries to match install
	settings to allow access to stuff in /proc.  More importantly, if
	mode is 755 then neither owner nor group are set.  This fixes bug
	1631136.  Added patch from haanjdj@xs4all.nl to fix an occasional
	core dump in m_decosf1.c.  This checks return code from task_threads.
	Made sure all get_system_info functions are declared void.  Fixed
	string termination bug.  Cleaned up documetation for sunos5.

Tue Aug  8 2006 - wnl (3.6.1)
	For Solaris, changed the tag "swap" to "total swap" to clarify
	what is beign displayed.  Note that the calculations are still the
	same: the display is just showing total rather than total - free.

Thu Apr 27 2006 - wnl (3.6)
	Added patches for linux-style sort shortcuts and for Unixware
	support in configure (patch 1474427).  Fixed sunos5 to do slow start
	and to ensure cpucount is set (patch 1477386).  Added pagination
	routines to display.c and modified show_help to use it, since the
	help screen is now longer than 24 lines.  Applied patch for unixware
	support that adds check for mas library (patch #1474423).  Solaris
	cpu percent now reflects a percentage of the entire server, rather
	than a single cpu (bug 1478138).

Mon Mar 27 2006 - wnl (3.6)
	The production release of version 3.6.  Fixed a minor scaling
	bug in the decosf1 module.  Support for MacOS X is officially
	withdrawn although the macosx module is still part of the
	distribution.  Hopefully this is a temporary situation.
	Documentation updated.


Wed Feb 15 2006 - wnl (3.6beta5)
	Minor changes to eliminate warnings from the Sun Studio compiler.
	These were mostly sloppy argument declarations.  I also added
	message.h to provide an interface file for just the message
	related functions of display.c.

Mon Dec 26 2005 - wnl (3.6beta4)
	Added new netbsd module, courtesy of Simon Burge.
	Fixed a few bugs in SVR4 module and added its use to
	configure.ac, thanks to Sanchet Dighe.  Also ensured that the
	novpath Makefile was in the distribution.
	Fixed portability problem in display.c


Mon Oct 24 2005 - wnl (3.6beta3)
	Set up a color tagging mechanism in color.c to allow for the
	dynamic creation of tag names to contol color highlighting.
	These names are partially derived from the tags used to label
	memory and swap information on the screen, thus are driven by
	the machine module itself.  Added -T option to list color 
	highlighting information.  Help screen now includes the actual
	list of sort order names.  Incorporated some minor fixes to
	the main code from the Freebsd source tree.  Fixed bug #1324582.
	Freebsd 5: removed WCPU column and added THR column.  Display
	for freebsd 4 and earlier unchanged since they don't track
	threads in the kernel.  Added LICENSE file to distribution.

Wed Oct 12 2005 - wnl (3.6beta2)
	Major overhaul to display.c.  All lines of the display are
	directly tracked and controlled via display_write and its
	companion display_fmt.  Added support for complete control
	of ANSI color on the screen: this will be used in the future
	to allow for full use of color everywhere on the screen.
	Signal handling code now uses sigaction on all systems that
	support it.  Restored the freebsd module and did away with
	freebsd4, and upgraded freebsd module to support 5.x.
	Fix bug #1306099 (wio(wait) timer ignored on OSF1).

Fri Sep 23 2005 - wnl (3.6beta1)
	Fixed bugs #1266341 (compilation errors with gcc 4.x),
	#1156464 (cpu% field for sunos), #1156243 (compilation
	errors on AIX).  Applied patches #1217855 (Solaris 10
	xarch flag).  Overhaul of sunos5 module, making code more
	efficient and easier to follow.  Got rid of need for MEMTYPE
	type in utils.h.  Changed all memory statistics data in the
	module specification from an int to a long to better support
	64-bit systems with lots of memory. Moved all unused modules
	out of the distribution (I will add them back in as needed).
	Moved freebsd module to freebsd4 as it won't work with 5.x
	(a new module will be necessary). Added support to configure
	for makes that don't understand VPATH. Updated documentation:
	man page, FAQ, README, INSTALL.

Mon Jan 24 2005 - wnl (3.6alpha10)
	Updated aix43 module with ANSI function declarations and fixed
	declaration of get_system_info.  Configure now uses irixsgi
	module for irix6* systems.  Updates to the following modules:
	irixsgi, sunos5.  Fixed null pointer bug in color.c. Removed
	some useless code and definitions in display.c


Sun Nov 28 2004 - wnl (3.6alpha9)
	Replace AIX 5 module with alternate (bug 1056565).
	Fixed vulnerability in use of snprintf.

Fri Oct 22 2004 - wnl (3.6alpha8)
	Support for linux 2.6, added more stuff to memory and swap lines.
	Updated linuxthr module, which is only useful on 2.4 and earlier.
	Added some color support back in (feature request 930588), but
	still need to add it back to the per-process display.  Added
	OSF 5 support (untested).
	Fixed bug 1017951 (invalid process count argument not caught)


old_modules/m_ncr3000.c
/*
 * top - a top users display for Unix
 *
 * SYNOPSIS:  For NCR 3000 series systems Release 2.00.02 and above -
 *	works on 2.03.00 and earlier (and probably later) OS releases.
 * 	(Intel based System V Release 4)
 *
 * DESCRIPTION:
 *      System V release 4     for NCR 3000 series OS Rel 02.03.00 and above
 *
 * LIBS:  -lelf
 *
 * AUTHORS:  Andrew Herbert     <andrew@werple.apana.org.au>
 *           Robert Boucher     <boucher@sofkin.ca>
 *           Jeff Janvrin	<jeff.janvrin@columbiasc.ncr.com>
 *                              did the port to statfs (2.03)
 */

#include "top.h"
#include "machine.h"
#include "utils.h"
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <dirent.h>
#include <nlist.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <sys/procfs.h>
#include <sys/sysinfo.h>
#include <sys/sysmacros.h>
#include <sys/vmmeter.h>
#include <vm/anon.h>
#include <sys/priocntl.h>
#include <sys/rtpriocntl.h>
#include <sys/tspriocntl.h>
#include <sys/procset.h>
#include <sys/var.h>

#define UNIX "/stand/unix"
#define KMEM "/dev/kmem"
#define PROCFS "/proc"
#define CPUSTATES	5

#ifndef PRIO_MAX
#define PRIO_MAX	20
#endif
#ifndef PRIO_MIN
#define PRIO_MIN	-20
#endif

#ifndef FSCALE
#define FSHIFT  8		/* bits to right of fixed binary point */
#define FSCALE  (1<<FSHIFT)
#endif

#define loaddouble(x) ((double)(x) / FSCALE)
#define percent_cpu(x) ((double)(x)->pr_cpu / FSCALE)
#define weighted_cpu(pct, pp) ( ((pp)->pr_time.tv_sec) == 0 ? 0.0 : \
        ((pp)->pr_cpu) / ((pp)->pr_time.tv_sec) )
#define pagetok(size) ctob(size) >> LOG1024

/* definitions for the index in the nlist array */
#define X_AVENRUN	0
#define X_MPID		1
#define X_V		2
#define X_NPROC		3
#define X_ANONINFO	4
#define X_TOTAL		5
#define X_SYSINFO	6

static struct nlist nlst[] =
{
{"avenrun"},			/* 0 */
{"mpid"},			/* 1 */
{"v"},			/* 2 */
{"nproc"},			/* 3 */
{"anoninfo"},			/* 4 */
{"total"},			/* 5 */
{"sysinfo"},			/* 6 */
{NULL}
};

static unsigned long avenrun_offset;
static unsigned long mpid_offset;
static unsigned long nproc_offset;
static unsigned long anoninfo_offset;
static unsigned long total_offset;
static unsigned long sysinfo_offset;

/* get_process_info passes back a handle.  This is what it looks like: */

struct handle
  {
    struct prpsinfo **next_proc;/* points to next valid proc pointer */
    int remaining;		/* number of pointers remaining */
  };

/*
 *  These definitions control the format of the per-process area
 */

static char header[] =
"  PID X        PRI NICE  SIZE   RES STATE   TIME   WCPU    CPU COMMAND";
/* 0123456   -- field to fill in starts at header+6 */
#define UNAME_START 6
#define Proc_format \
	"%5d %-8.8s %3d %4d %5s %5s %-5s %6s %3d.0%% %5.2f%% %.16s"

char *state_abbrev[] =
{"", "sleep", "run", "zombie", "stop", "start", "cpu", "swap"};

int process_states[8];
char *procstatenames[] =
{
  "", " sleeping, ", " running, ", " zombie, ", " stopped, ",
  " starting, ", " on cpu, ", " swapped, ",
  NULL
};

int cpu_states[CPUSTATES];
char *cpustatenames[] =
{"idle", "user", "kernel", "wait", "swap", NULL};

/* these are for detailing the memory statistics */

int memory_stats[5];
char *memorynames[] =
{"K real, ", "K active, ", "K free, ", "K swap, ", "K free swap", NULL};

static int kmem = -1;
static int nproc;
static int bytes;
static int use_stats = 0;
static struct prpsinfo *pbase;
static struct prpsinfo **pref;
static DIR *procdir;

/* useful externals */
extern int errno;
extern char *sys_errlist[];
extern char *myname;
extern int check_nlist ();
extern int getkval ();
extern void perror ();
extern void getptable ();
extern void quit ();
extern int nlist ();

int
machine_init (struct statics *statics)
  {
    static struct var v;

    /* fill in the statics information */
    statics->procstate_names = procstatenames;
    statics->cpustate_names = cpustatenames;
    statics->memory_names = memorynames;

    /* get the list of symbols we want to access in the kernel */
    if (nlist (UNIX, nlst))
      {
	(void) fprintf (stderr, "Unable to nlist %s\n", UNIX);
	return (-1);
      }

    /* make sure they were all found */
    if (check_nlist (nlst) > 0)
      return (-1);

    /* open kernel memory */
    if ((kmem = open (KMEM, O_RDONLY)) == -1)
      {
	perror (KMEM);
	return (-1);
      }

    /* get the symbol values out of kmem */
    /* NPROC Tuning parameter for max number of processes */
    (void) getkval (nlst[X_V].n_value, (int *) &v, sizeof (struct var), nlst[X_V].n_name);
    nproc = v.v_proc;

    /* stash away certain offsets for later use */
    mpid_offset = nlst[X_MPID].n_value;
    nproc_offset = nlst[X_NPROC].n_value;
    avenrun_offset = nlst[X_AVENRUN].n_value;
    anoninfo_offset = nlst[X_ANONINFO].n_value;
    total_offset = nlst[X_TOTAL].n_value;
/* JJ this may need to be changed */
    sysinfo_offset = nlst[X_SYSINFO].n_value;

    /* allocate space for proc structure array and array of pointers */
    bytes = nproc * sizeof (struct prpsinfo);
    pbase = (struct prpsinfo *) malloc (bytes);
    pref = (struct prpsinfo **) malloc (nproc * sizeof (struct prpsinfo *));

    /* Just in case ... */
    if (pbase == (struct prpsinfo *) NULL || pref == (struct prpsinfo **) NULL)
      {
	(void) fprintf (stderr, "%s: can't allocate sufficient memory\n", myname);
	return (-1);
      }

    if (!(procdir = opendir (PROCFS)))
      {
	(void) fprintf (stderr, "Unable to open %s\n", PROCFS);
	return (-1);
      }

    if (chdir (PROCFS))
      {				/* handy for later on when we're reading it */
	(void) fprintf (stderr, "Unable to chdir to %s\n", PROCFS);
	return (-1);
      }

    /* all done! */
    return (0);
  }

char *
format_header (char *uname_field)
{
  register char *ptr;

  ptr = header + UNAME_START;
  while (*uname_field != '\0')
    *ptr++ = *uname_field++;

  return (header);
}

void
get_system_info (struct system_info *si)
{
  long avenrun[3];
  struct sysinfo sysinfo;
  static struct sysinfo *mpinfo = NULL;	/* array, per-processor sysinfo structures. */
  struct vmtotal total;
  struct anoninfo anoninfo;
  static time_t cp_old[CPUSTATES];
  static time_t cp_diff[CPUSTATES];	/* for cpu state percentages */
  static int num_cpus;
  static int fd_cpu = 0;
  register int i;

  if ( use_stats == 1) {
    if ( fd_cpu == 0 ) {
      if ((fd_cpu = open("/stats/cpuinfo", O_RDONLY)) == -1) {
        (void) fprintf (stderr, "%s: Open of /stats/cpuinfo failed\n", myname);
	quit(2);
      }
      if (read(fd_cpu, &num_cpus, sizeof(int)) != sizeof(int)) {
        (void) fprintf (stderr, "%s: Read of /stats/cpuinfo failed\n", myname);
	quit(2);
      }
      close(fd_cpu);
    }
    if (mpinfo == NULL) {
      mpinfo = (struct sysinfo *)calloc(num_cpus, sizeof(mpinfo[0]));
      if (mpinfo == NULL) {
        (void) fprintf (stderr, "%s: can't allocate space for per-processor sysinfos\n", myname);
        quit(12);
      }
    }
    /* Read the per cpu sysinfo structures into mpinfo struct. */
    read_sysinfos(num_cpus, mpinfo);
    /* Add up all of the percpu sysinfos to get global sysinfo */
    sysinfo_data(num_cpus, &sysinfo, mpinfo);
  } else {
    (void) getkval (sysinfo_offset, &sysinfo, sizeof (struct sysinfo), "sysinfo");
  }

  /* convert cp_time counts to percentages */
  (void) percentages (CPUSTATES, cpu_states, sysinfo.cpu, cp_old, cp_diff);

  /* get mpid -- process id of last process */
  (void) getkval (mpid_offset, &(si->last_pid), sizeof (si->last_pid),
		  "mpid");

  /* get load average array */
  (void) getkval (avenrun_offset, (int *) avenrun, sizeof (avenrun), "avenrun");

  /* convert load averages to doubles */
  for (i = 0; i < 3; i++)
    si->load_avg[i] = loaddouble (avenrun[i]);

  /* get total -- systemwide main memory usage structure */
  (void) getkval (total_offset, (int *) (&total), sizeof (total), "total");
  /* convert memory stats to Kbytes */
  memory_stats[0] = pagetok (total.t_rm);
  memory_stats[1] = pagetok (total.t_arm);
  memory_stats[2] = pagetok (total.t_free);
  (void) getkval (anoninfo_offset, (int *) (&anoninfo), sizeof (anoninfo),
		  "anoninfo");
  memory_stats[3] = pagetok (anoninfo.ani_max - anoninfo.ani_free);
  memory_stats[4] = pagetok (anoninfo.ani_max - anoninfo.ani_resv);

  /* set arrays and strings */
  si->cpustates = cpu_states;
  si->memory = memory_stats;
}

static struct handle handle;

caddr_t
get_process_info (
		   struct system_info *si,
		   struct process_select *sel,
		   int (*compare) ())
{
  register int i;
  register int total_procs;
  register int active_procs;
  register struct prpsinfo **prefp;
  register struct prpsinfo *pp;

  /* these are copied out of sel for speed */
  int show_idle;
  int show_system;
  int show_uid;

  /* Get current number of processes */
  (void) getkval (nproc_offset, (int *) (&nproc), sizeof (nproc), "nproc");

  /* read all the proc structures */
  getptable (pbase);

  /* get a pointer to the states summary array */
  si->procstates = process_states;

  /* set up flags which define what we are going to select */
  show_idle = sel->idle;
  show_system = sel->system;
  show_uid = sel->uid != -1;

  /* count up process states and get pointers to interesting procs */
  total_procs = 0;
  active_procs = 0;
  (void) memset (process_states, 0, sizeof (process_states));
  prefp = pref;

  for (pp = pbase, i = 0; i < nproc; pp++, i++)
    {
      /*
	 *  Place pointers to each valid proc structure in pref[].
	 *  Process slots that are actually in use have a non-zero
	 *  status field.  Processes with SSYS set are system
	 *  processes---these get ignored unless show_sysprocs is set.
	 */
      if (pp->pr_state != 0 &&
	  (show_system || ((pp->pr_flag & SSYS) == 0)))
	{
	  total_procs++;
	  process_states[pp->pr_state]++;
	  if ((!pp->pr_zomb) &&
	      (show_idle || (pp->pr_state == SRUN) || (pp->pr_state == SONPROC)) &&
	      (!show_uid || pp->pr_uid == (uid_t) sel->uid))
	    {
	      *prefp++ = pp;
	      active_procs++;
	    }
	}
    }

  /* if requested, sort the "interesting" processes */
  if (compare != NULL)
      qsort ((char *) pref, active_procs, sizeof (struct prpsinfo *), compare);

  /* remember active and total counts */
  si->p_total = total_procs;
  si->p_active = active_procs;

  /* pass back a handle */
  handle.next_proc = pref;
  handle.remaining = active_procs;
  return ((caddr_t) & handle);
}

char fmt[MAX_COLS];			/* static area where result is built */

char *
format_next_process (
		      caddr_t handle,
		      char *(*get_userid) ())
{
  register struct prpsinfo *pp;
  struct handle *hp;
  register long cputime;
  register double pctcpu;

  /* find and remember the next proc structure */
  hp = (struct handle *) handle;
  pp = *(hp->next_proc++);
  hp->remaining--;

  /* get the cpu usage and calculate the cpu percentages */
  cputime = pp->pr_time.tv_sec;
  pctcpu = percent_cpu (pp);

  /* format this entry */
  (void) sprintf (fmt,
		  Proc_format,
		  pp->pr_pid,
		  (*get_userid) (pp->pr_uid),
		  pp->pr_pri - PZERO,
		  pp->pr_nice - NZERO,
		  format_k(pagetok (pp->pr_size)),
		  format_k(pagetok (pp->pr_rssize)),
		  state_abbrev[pp->pr_state],
		  format_time(cputime),
		  (pp->pr_cpu & 0377),
		  100.0 * pctcpu,
		  pp->pr_fname);

  /* return the result */
  return (fmt);
}

/*
 * check_nlist(nlst) - checks the nlist to see if any symbols were not
 *		found.  For every symbol that was not found, a one-line
 *		message is printed to stderr.  The routine returns the
 *		number of symbols NOT found.
 */
int
check_nlist (register struct nlist *nlst)
{
  register int i;
  struct stat stat_buf;

  /* check to see if we got ALL the symbols we requested */
  /* this will write one line to stderr for every symbol not found */

  i = 0;
  while (nlst->n_name != NULL)
    {
      if (nlst->n_type == 0)
	{
	  if (strcmp("sysinfo", nlst->n_name) == 0)
	    {
		/* check to see if /stats file system exists. If so, 	*/
		/* ignore error. 					*/
		if ( !((stat("/stats/sysinfo", &stat_buf) == 0) && 
		  (stat_buf.st_mode & S_IFREG)) )
		  {
		    (void) fprintf (stderr, "kernel: no symbol named `%s'\n", nlst->n_name);
		    i = 1;
		  } else {
		    use_stats = 1;
		  }
	    } else {

	      /* this one wasn't found */
	      (void) fprintf (stderr, "kernel: no symbol named `%s'\n", nlst->n_name);
	      i = 1;
	    }
	}
      nlst++;
    }
  return (i);
}


/*
 *  getkval(offset, ptr, size, refstr) - get a value out of the kernel.
 *	"offset" is the byte offset into the kernel for the desired value,
 *  	"ptr" points to a buffer into which the value is retrieved,
 *  	"size" is the size of the buffer (and the object to retrieve),
 *  	"refstr" is a reference string used when printing error meessages,
 *	    if "refstr" starts with a '!', then a failure on read will not
 *  	    be fatal (this may seem like a silly way to do things, but I
 *  	    really didn't want the overhead of another argument).
 *
 */
int
getkval (
	  unsigned long offset,
	  int *ptr,
	  int size,
	  char *refstr)
{
  if (lseek (kmem, (long) offset, 0) == -1)
    {
      if (*refstr == '!')
	refstr++;
      (void) fprintf (stderr, "%s: lseek to %s: %s\n",
		      myname, refstr, sys_errlist[errno]);
      quit (22);
    }
  if (read (kmem, (char *) ptr, size) == -1)
    if (*refstr == '!')
      /* we lost the race with the kernel, process isn't in memory */
      return (0);
    else
      {
	(void) fprintf (stderr, "%s: reading %s: %s\n",
			myname, refstr, sys_errlist[errno]);
	quit (23);
      }
  return (1);
}

/* comparison routine for qsort */

/*
 *  proc_compare - comparison function for "qsort"
 *	Compares the resource consumption of two processes using five
 *  	distinct keys.  The keys (in descending order of importance) are:
 *  	percent cpu, cpu ticks, state, resident set size, total virtual
 *  	memory usage.  The process states are ordered as follows (from least
 *  	to most important):  WAIT, zombie, sleep, stop, start, run.  The
 *  	array declaration below maps a process state index into a number
 *  	that reflects this ordering.
 */


unsigned char sorted_state[] =
{
  0,				/* not used		*/
  3,				/* sleep		*/
  6,				/* run			*/
  2,				/* zombie		*/
  4,				/* stop			*/
  5,				/* start		*/
  7,				/* run on a processor   */
  1				/* being swapped (WAIT)	*/
};

int
proc_compare (
	       struct prpsinfo **pp1,
	       struct prpsinfo **pp2)
{
    register struct prpsinfo *p1;
    register struct prpsinfo *p2;
    register long result;

    /* remove one level of indirection */
    p1 = *pp1;
    p2 = *pp2;

    /* compare percent cpu (pctcpu) */
    if ((result = (long) (p2->pr_cpu - p1->pr_cpu)) == 0)
      {
	/* use cpticks to break the tie */
	if ((result = p2->pr_time.tv_sec - p1->pr_time.tv_sec) == 0)
	  {
	    /* use process state to break the tie */
	    if ((result = (long) (sorted_state[p2->pr_state] -
				  sorted_state[p1->pr_state])) == 0)
	      {
		/* use priority to break the tie */
		if ((result = p2->pr_oldpri - p1->pr_oldpri) == 0)
		  {
		    /* use resident set size (rssize) to break the tie */
		    if ((result = p2->pr_rssize - p1->pr_rssize) == 0)
		      {
			/* use total memory to break the tie */
			result = (p2->pr_size - p1->pr_size);
		      }
		  }
	      }
	  }
      }
    return (result);
  }

/*
get process table
*/
void
getptable (struct prpsinfo *baseptr)
{
  struct prpsinfo *currproc;	/* pointer to current proc structure	*/
  int numprocs = 0;
  struct dirent *direntp;

  for (rewinddir (procdir); direntp = readdir (procdir);)
    {
      int fd;

      if ((fd = open (direntp->d_name, O_RDONLY)) < 0)
	continue;

      currproc = &baseptr[numprocs];
      if (ioctl (fd, PIOCPSINFO, currproc) < 0)
	{
	  (void) close (fd);
	  continue;
	}

      numprocs++;
      (void) close (fd);
    }

  if (nproc != numprocs)
    nproc = numprocs;
}

/* return the owner of the specified process, for use in commands.c as we're
   running setuid root */
uid_t
proc_owner (pid_t pid)
{
  register struct prpsinfo *p;
  int i;
  for (i = 0, p = pbase; i < nproc; i++, p++)
    if (p->pr_pid == pid)
      return (p->pr_uid);

  return (-1);
}

int
setpriority (int dummy, int who, int niceval)
{
  int scale;
  int prio;
  pcinfo_t pcinfo;
  pcparms_t pcparms;
  tsparms_t *tsparms;

  strcpy (pcinfo.pc_clname, "TS");
  if (priocntl (0, 0, PC_GETCID, (caddr_t) & pcinfo) == -1)
    return (-1);

  prio = niceval;
  if (prio > PRIO_MAX)
    prio = PRIO_MAX;
  else if (prio < PRIO_MIN)
    prio = PRIO_MIN;

  tsparms = (tsparms_t *) pcparms.pc_clparms;
  scale = ((tsinfo_t *) pcinfo.pc_clinfo)->ts_maxupri;
  tsparms->ts_uprilim = tsparms->ts_upri = -(scale * prio) / 20;
  pcparms.pc_cid = pcinfo.pc_cid;

  if (priocntl (P_PID, who, PC_SETPARMS, (caddr_t) & pcparms) == -1)
    return (-1);

  return (0);
}

/****************************************************************
 * read_sysinfos() -						*
 *	Read all of the CPU specific sysinfo sturctures in from	*
 *	the /stats file system.					*
 ****************************************************************/
read_sysinfos(num_cpus, buf)
	int num_cpus;
	struct sysinfo	*buf;
{

	static int	fd1=0;	/* file descriptor for /stats/sysinfo */
	int		read_sz;

	/* Open /stats/sysinfo one time only and leave it open */
	if (fd1==0) { 
		if ((fd1 = open("/stats/sysinfo", O_RDONLY)) == -1)
			(void) fprintf (stderr, "%s: Open of /stats/sysinfo failed\n", myname);
	}
	/* reset the read pointer to the beginning of the file */
	if (lseek(fd1, 0L, SEEK_SET) == -1)
		(void) fprintf (stderr, "%s: lseek to beginning of /stats/sysinfo failed\n", myname);
	read_sz = num_cpus * sizeof(buf[0]);
	if (read(fd1, buf, read_sz) != read_sz)
		(void) fprintf (stderr, "%s: Read of /stats/sysinfo failed\n", myname);
}

/****************************************************************
 * sysinfo_data() -						*
 *	Add up all of the CPU specific sysinfo sturctures to	*
 *	make the GLOBAL sysinfo.				*
 ****************************************************************/
sysinfo_data(num_cpus, global_si, percpu_si)
	int num_cpus;
	struct sysinfo	*global_si;
	struct sysinfo	*percpu_si;
{
	struct sysinfo	*percpu_p;
	int		cpu, i, *global, *src;

	/* null out the global statistics from last sample */
	memset(global_si, 0, sizeof(struct sysinfo));

	percpu_p = (struct sysinfo *)percpu_si;
	for(cpu = 0; cpu < num_cpus; cpu++) {
		global = (int *)global_si;
		src = (int *)percpu_p;

		/* assume sysinfo ends on an int boundary */
		/* Currently, all of the struct sysinfo members are the same
		 * size as an int. If that changes, we may not be able to
		 * do this. But this should be safe.
		 */
		for(i=0; i<sizeof(struct sysinfo)/sizeof(int); i++) {
			*global++ += *src++;
		}
		percpu_p++;
	}
}
